### PR TITLE
Implement constant folding for CAST

### LINKED ISSUE #417: Implement constant folding for CAST
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
When pushing predicates down into various storage systems, it is common for them to support single column predicates of the form

```
const OP column
```

SQL allows users to write predicates against columns using whatever format they deem useful, and inserts the appropriate `CAST` functions where needed.

For example

```
echo "1" > /tmp/foo.csv
cargo run -p datafusion-cli
```
And then run:
```
> CREATE EXTERNAL TABLE foo(ts double)
STORED AS CSV
LOCATION '/tmp/foo.csv';
0 rows in set. Query took 0 seconds.
> explain verbose select * from foo where ts < 5;
+-----------------------------------------+--------------------------------------------------------------------------+
| plan_type                               | plan                                                                     |
+-----------------------------------------+--------------------------------------------------------------------------+
| logical_plan                            | Projection: #ts                                                          |
|                                         |   Filter: #ts Lt Int64(5)                                                |
|                                         |     TableScan: foo projection=None                                       |
| logical_plan after projection_push_down | Projection: #ts                                                          |
|                                         |   Filter: #ts Lt Int64(5)                                                |
|                                         |     TableScan: foo projection=Some([0])                                  |
| logical_plan after projection_push_down | Projection: #ts                                                          |
|                                         |   Filter: #ts Lt Int64(5)                                                |
|                                         |     TableScan: foo projection=Some([0])                                  |
| physical_plan                           | ProjectionExec: expr=[ts]                                                |
|                                         |   FilterExec: ts < CAST(5 AS Float64)                                    |
|                                         |     CsvExec: source=Path(/tmp/foo.csv: [/tmp/foo.csv]), has_header=false |
+-----------------------------------------+--------------------------------------------------------------------------+
4 rows in set. Query took 0 seconds.
```

You can see that the filter that was pushed down was `ts < CAST(5 AS Float64)  ` which is correct, but it will likely not be recognized by some storage systems (such as IOx).

**Describe the solution you'd like**
What I would like to see is the expression rewriter pass rewrite `ts < CAST(5 AS Float64)` into `ts < 5.0` (aka do the cast at plan time)

### PR DESCRIPTION
**Pull request by msathis**
# Which issue does this PR close?
Closes #417.

 # Rationale for this change
Optimize cast function during planning stage

# What changes are included in this PR?
Optimize cast function during planning stage

# Are there any user-facing changes?
N/A

### CODE REVIEW

**Review by houqp — APPROVED** [review]

File: datafusion/src/optimizer/constant_folding.rs [hunk 1]
```diff
@@ -247,6 +248,25 @@ impl<'a> ExprRewriter for ConstantRewriter<'a> {
                     }
                 }
             }
+            Expr::Cast {
+                expr: inner,
+                data_type,
+            } => match inner.as_ref() {
+                Expr::Literal(val) => {
+                    let scalar_array = val.to_array();
```
Comment (Dandandan) [inline datafusion/src/optimizer/constant_folding.rs:256]: Great idea👍 I think converting to an array is the best we can do for now, without duplicating the code.

**Review by Dandandan — APPROVED** [review]
I think this is looking superb! Thanks @msathis

File: datafusion/src/optimizer/constant_folding.rs [hunk 2]
```diff
@@ -724,6 +744,44 @@ mod tests {
...
+            expr: Box::new(Expr::Literal(ScalarValue::Utf8(Some("0".to_string())))),
+            data_type: DataType::Int32,
+        }];
+        let plan = LogicalPlanBuilder::from(&table_scan)
+            .project(proj)
+            .unwrap()
+            .build()
+            .unwrap();
+
+        let expected = "Projection: Int32(0)\
```
Comment (alamb) [inline datafusion/src/optimizer/constant_folding.rs:760]: this is beautiful ❤️  -- thank you @msathis

File: datafusion/src/optimizer/constant_folding.rs [hunk 3]
```diff
@@ -724,6 +744,44 @@ mod tests {
...
+            expr: Box::new(Expr::Literal(ScalarValue::Utf8(Some("".to_string())))),
+            data_type: DataType::Int32,
+        }];
+        let plan = LogicalPlanBuilder::from(&table_scan)
+            .project(proj)
+            .unwrap()
+            .build()
+            .unwrap();
+
+        let expected = "Projection: Int32(NULL)\
```
Comment (alamb) [inline datafusion/src/optimizer/constant_folding.rs:779]: 👍

**Review by alamb — COMMENTED** [review]
Thanks @msathis !

**Pull request merged**